docs/RFCS/20210722_on_update_and_row_rehoming.md

```diff
@@ -0,0 +1,250 @@
+```
+
+Concretely, creating a table with a `quantity_on_hand` column that applies a
+value of 50 ON UPDATE would look like:
+
+```sql
+CREATE TABLE inventories (
+  product_id        INT,
+  warehouse_id      INT,
+  quantity_on_hand  INT ON UPDATE 50,
+  PRIMARY KEY (product_id, warehouse_id)
+);
+```
+
+This syntax also applies to `ALTER TABLE ALTER COLUMN`:
+
+```sql
+-- Modifying an existing ON UPDATE expression or adding a new one
+ALTER TABLE <table_name> ALTER COLUMN <column_name> SET ON UPDATE <update_expr>
+-- Dropping an ON UPDATE expression
+ALTER TABLE <table_name> ALTER COLUMN <column_name> DROP ON UPDATE
+```
+
+Concretely with our inventories example, we have:
+
+```sql
+-- Modifying the existing ON UPDATE expression to 50
+ALTER TABLE inventories ALTER COLUMN quantity_on_hand SET ON UPDATE 50
+-- Dropping an ON UPDATE expression
+ALTER TABLE inventories ALTER COLUMN quantity_on_hand DROP ON UPDATE
+```
+
+#### Example Usage
+
+```sql
+CREATE TABLE inventories (
+    product_id        INT,
+    warehouse_id      INT,
+    quantity_on_hand  INT,
+    PRIMARY KEY (product_id, warehouse_id)
+  );
+INSERT INTO inventories (product_id, warehouse_id, quantity_on_hand) VALUES (1, 1, 1);
+SELECT quantity_on_hand FROM inventories;
+> 1
+UPDATE inventories SET product_id = 2 WHERE warehouse_id = 1;
+SELECT quantity_on_hand FROM inventories;
+> 1
+ALTER TABLE inventories ALTER COLUMN quantity_on_hand SET ON UPDATE 50;
+UPDATE inventories SET product_id = 3 WHERE warehouse_id = 1;
+SELECT quantity_on_hand FROM inventories;
+> 50
+UPDATE inventories SET quantity_on_hand = 100;
+-- how to UPDATE keeping current value for quantity
+UPDATE inventories i SET (product_id, quantity_on_hand) = (4, i.quantity_on_hand);
+SELECT quantity_on_hand FROM inventories;
+> 100
+```
+
+## Row Rehoming
+
+To implement row rehoming, an ON UPDATE constraint will be added to the
+crdb_region column in REGIONAL BY ROW tables. This will make it so that when an
+UPDATE occurs that does not touch `crdb_region`, the `gateway_region()` default
+clause will be evaluated and applied.
+
+This has the potential to cause thrashing if UPDATEs happen from many different
+regions, so this behavior will be gated behind a cluster setting,
+`sql.defaults.auto_rehome.enabled`.
+
+If the user wishes to apply this behavior to just one of their tables as opposed
+to applying a cluster-wide setting, they can create the REGIONAL BY ROW table
+and alter its `crdb_region` column with an ON UPDATE expression , or they can specify
+ON UPDATE on their `crdb_internal_region` column and specify the column with
+REGIONAL BY ROW AS.
+
+Example of creating an auto-rehoming table:
+
+```sql
+CREATE TABLE test (
+  p int,
+  region crdb_internal_region
+    DEFAULT default_to_database_primary_region(gateway_region())
+    ON UPDATE default_to_database_primary_region(gateway_region())
+)
+  LOCALITY REGIONAL BY ROW AS region;
+```
+
+## Drawbacks
```

Comment:
There's another potential drawback, though I think any implementation would suffer from it. Might be worth documenting, however.

The region column is part of the primary key, so the primary key may be mutating for every UPDATE. This is a drawback because it will require all column families that have an indexed column to be fetched for every UPDATE. 

Here's an example:

```
statement ok
CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT,
  b INT,
  INDEX a_idx (a),
  INDEX b_idx (b),
  FAMILY (k, a),
  FAMILY (b)
)

statement ok
INSERT INTO t VALUES (10, 20, 30)

# We do not need to scan the column family with b because b_idx will never
# change.
query T kvtrace
UPDATE t SET a = 220 WHERE k = 10
----
Scan /Table/53/1/10/0
Put /Table/53/1/10/0 -> /TUPLE/2:2:Int/220
Del /Table/53/2/20/10/0
CPut /Table/53/2/220/10/0 -> /BYTES/ (expecting does not exist)

# We must scan the column family with b because b_idx could change, if the PK
# changes.
query T kvtrace
UPDATE t SET a = 222, k = 10 WHERE k = 10
----
Scan /Table/53/1/10{-/#}
Scan /Table/53/1/10/{1/1-#}
Put /Table/53/1/10/0 -> /TUPLE/2:2:Int/222
Del /Table/53/2/220/10/0
CPut /Table/53/2/222/10/0 -> /BYTES/ (expecting does not exist)
```

This may hurt performance for customers who put large, indexed columns, like big JSON blobs, in separate column families to optimize their workloads.

Reply:
Ah good point thanks. I don't know of any way to get around this, so I'll add it as a drawback for all options.